### PR TITLE
Bug fix for publishing modules with RequiredModules specified in .psd1

### PR DESCRIPTION
**Pull request by alerickson**
# PR Summary
Currently publishing modules that specify `RequiredModules` in the module manifest fails.  This is due to the fact that `Publish-PSResource` was completed before `FindHelper` was complete, so calls to search for dependency packages in a repository were not available.  This PR properly calls into `FindHelper` to search for the specified dependency modules. 

Error handling has also been updated a bit since there were some bugs found there as well.

## PR Context

Closes issue #486, #630, and  #489 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: src/code/PublishPSResource.cs [hunk 1]
```diff
@@ -160,9 +161,10 @@ public string DestinationPath
 
         #region Members
 
+        private CancellationToken _cancellationToken;
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs:165]: Please include in the PR description, a short summary of the issue and how these changes address the problem.

File: src/code/PublishPSResource.cs [hunk 2]
```diff
@@ -792,15 +804,27 @@ private bool CheckDependenciesExist(Hashtable dependencies, string repositoryUri
...
+                    this.ThrowTerminatingError(new ErrorRecord(
+                        new ArgumentException("Dependency version is not in the proper format"),
+                        "IncorrectVersionFormat",
+                        ErrorCategory.InvalidArgument,
+                        this));
+                }
+
+                bool depPrerelease = depVersion.Contains("-") ? true : false;
+
                 var repository = new[] { repositoryUri };
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]: The caller now passes in `repository.Name` to this method, so the parameter name should reflect that.

File: src/code/PublishPSResource.cs [hunk 3]
```diff
@@ -433,19 +436,28 @@ private bool IsValidModuleManifest(string moduleManifestPath)
                 if (pwsh.HadErrors)
                 {
                     var message = string.Empty;
-                    if (string.IsNullOrWhiteSpace((results[0].BaseObject as PSModuleInfo).Author))
-                    {
-                        message = "No author was provided in the module manifest. The module manifest must specify a version, author and description.";
-                    }
-                    else if (string.IsNullOrWhiteSpace((results[0].BaseObject as PSModuleInfo).Description))
+
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs:452]: pwsh.Invoke() above will throw if the command it is running throws.  So it should be wrapped in a try/catch to be safe.

File: src/code/PublishPSResource.cs [hunk 4]
```diff
@@ -433,19 +436,28 @@ private bool IsValidModuleManifest(string moduleManifestPath)
...
+                            message = "No description was provided in the module manifest. The module manifest must specify a version, author and description. Run 'Test-ModuleManifest' to validate the file.";
+                        }
+                        else if (string.IsNullOrWhiteSpace((results[0].BaseObject as PSModuleInfo).Version.ToString()))
+                        {
+                            message = "No verison was provided in the module manifest. The module manifest must specify a version, author and description. Run 'Test-ModuleManifest' to validate the file.";
+                        }
                     }
-                    else
+
+                    if (string.IsNullOrEmpty(message))
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]:
```suggestion
                    if (string.IsNullOrEmpty(message) && pwsh.Streams.Error.Count > 0)

```

File: src/code/PublishPSResource.cs [hunk 5]
```diff
@@ -433,19 +436,28 @@ private bool IsValidModuleManifest(string moduleManifestPath)
...
                     {
                         // This will handle version errors
                         var error = pwsh.Streams.Error;
-                        message = error[0].ToString();
+                        message = error[0].ToString() + "Run 'Test-ModuleManifest' to validate the module manifest.";
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]:
```suggestion
                        message = pwsh.Streams.Error[0].ToString() + "Run 'Test-ModuleManifest' to validate the module manifest.";

```

File: src/code/PublishPSResource.cs [hunk 6]
```diff
@@ -792,15 +804,27 @@ private bool CheckDependenciesExist(Hashtable dependencies, string repositoryUri
...
+                if (!Utils.TryParseVersionOrVersionRange(depVersion, out versionRange))
+                {
+                    this.ThrowTerminatingError(new ErrorRecord(
+                        new ArgumentException("Dependency version is not in the proper format"),
+                        "IncorrectVersionFormat",
+                        ErrorCategory.InvalidArgument,
+                        this));
+                }
+
+                bool depPrerelease = depVersion.Contains("-") ? true : false;
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]: This local variable does not appear to be used anywhere.

File: test/PublishPSResource.Tests.ps1 [hunk 7]
```diff
@@ -74,32 +74,29 @@ Describe "Test Publish-PSResource" {
...
 
     It "Publish a module with a dependency that is not published, should throw" {
         $version = "1.0.0"
         $dependencyVersion = "2.0.0"
-        New-ModuleManifest -Path (Join-Path -Path $script:PublishModuleBase -ChildPath "$script:PublishModuleName.psd1") -ModuleVersion $version -Description "$script:PublishModuleName module" -RequiredModules @(@{ModuleName="PackageManagement"; ModuleVersion="$dependencyVersion"})
-
-        Publish-PSResource -Path $script:PublishModuleBase -ErrorAction SilentlyContinue
+        New-ModuleManifest -Path (Join-Path -Path $script:PublishModuleBase -ChildPath "$script:PublishModuleName.psd1") -ModuleVersion $version -Description "$script:PublishModuleName module" -RequiredModules @(@{ModuleName = 'PackageManagement'; ModuleVersion = '1.4.4' })
 
+        Publish-PSResource -Path $script:PublishModuleBase -Verbose -ErrorAction SilentlyContinue
```
Comment (PaulHigin) [inline test/PublishPSResource.Tests.ps1]: If this throws, then you should use the Pester `-Should throw` here.

File: test/PublishPSResource.Tests.ps1 [hunk 8]
```diff
@@ -74,32 +74,29 @@ Describe "Test Publish-PSResource" {
...
         $Error[0].FullyQualifiedErrorId | Should -be "DependencyNotFound,Microsoft.PowerShell.PowerShellGet.Cmdlets.PublishPSResource"
```
Comment (PaulHigin) [inline test/PublishPSResource.Tests.ps1]: Please add tests for 'invalid manifest' and for 'invalid version format'.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/code/PublishPSResource.cs [hunk 9]
```diff
@@ -792,19 +816,31 @@ private bool CheckDependenciesExist(Hashtable dependencies, string repositoryUri
...
+                        this));
+                }
+                
                 // Search for and return the dependency if it's in the repository.
-                // TODO: When find is complete, uncomment beginFindHelper method below  (resourceNameParameterHelper)
-                //var dependencyFound = findHelper.beginFindHelper(depName, type, depVersion, true, null, null, repository, Credential, false, false);
-                // TODO: update the type from PSObject to PSResourceInfo
-                List<PSObject> dependencyFound = null;
+                FindHelper findHelper = new FindHelper(_cancellationToken, this);
+                bool depPrerelease = depVersion.Contains("-") ? true : false;
```
Comment (anamnavi) [inline src/code/PublishPSResource.cs]:
```suggestion
                bool depPrerelease = depVersion.Contains("-");

```

**Review by anamnavi — COMMENTED** [review]
LGTM, minor suggestion for variable assignment.

This will probably resolve in merge conflicts with the Publish scripts PR feel free to ping and we can incorporate both changes together

**Review by anamnavi — APPROVED** [review]

**Pull request merged**